File: web/page.ts
/**
 * Composing the page from the acts.
 *
 * Every section is written into the document that ships, so the page is
 * complete and readable before a single line of script runs. Part 2 is then
 * rebuilt in place when the reader switches profile.
 */

import { ACTS, perProfile, type Act, type Context } from "./acts/index";
import { PARTS } from "./copy/units";
import type { Payload, Profile, UserId } from "./types/index";
import { profile as profileOf } from "./document";

/**
 * The line handing the reader to the next act.
 *
 * It is appended to the act body, not to the section around it: the profile
 * switch replaces the whole body with what it built, so a bridge added
 * outside would survive the first paint and vanish on the first swap.
 */
function bridge(act: Act): string {
  return act.next ? `<p class="act-next">${act.next}</p>` : "";
}

/** One act's heading and body, as the shell's `<section>` expects them. */
export function section(act: Act, body: string): string {
  return '<header class="act-head">'
    + `<p class="eyebrow">${act.eyebrow}</p>`
    + `<h2 class="act-title">${act.title}</h2></header>`
    + `<div class="act-body">${body}</div>`;
}

/** Twelve acts in uppercase mono, grouped into the three parts. */
export function rail(): string {
  let out = "";
  let part: number | null = null;
  for (const act of ACTS) {
    if (act.part !== part) {
      if (part !== null) out += "</ol>";
      out += `<p class="rail-part">${PARTS[act.part - 1]}</p><ol>`;
      part = act.part;
    }
    out += `<li><a href="#act-${act.id}" data-rail="${act.id}">`
      + `<span class="num">${act.id}</span>${act.title}</a></li>`;
  }
  return `${out}</ol>`;
}

function context(payload: Payload, user: UserId, profile: Profile): Context {
  return {
    payload, user, profile,
    selection: { user, week: profile.defaultWeek, day: profile.defaultDay },
  };
}

/** Every act body for one profile, keyed by act id. */
export function bodies(payload: Payload, user: UserId): Record<string, string> {
  const profile = profileOf(payload, user);
  const ctx = context(payload, user, profile);
  const out: Record<string, string> = {};
  for (const act of ACTS) {
    out[act.id] = act.build(ctx) + bridge(act);
  }
  return out;
}

/** The act bodies the profile switch replaces: part 2 and nothing else. */
export const PART_TWO = ACTS.filter(perProfile).map((a) => a.id);

File: render/page.py
"""Compose `docs/index.html` from the hand-written shell and the acts.

Parts 1 and 3 are rendered once. Part 2 is rendered for every profile: the
default profile's copy goes into the page so it is complete before any script
runs, and both go into the payload so the switch has something to swap in.
"""

from __future__ import annotations

from pathlib import Path

from copytext import STRINGS, t

from .acts import ACTS, Context

SHELL = Path(__file__).parent.parent / "site" / "index.html"


def section(act, body: str) -> str:
    return (f'<header class="act-head"><p class="eyebrow">{t(act.eyebrow)}</p>'
            f'<h2 class="act-title">{t(act.title)}</h2></header>'
            f'<div class="act-body">{body}</div>')


def bridge(act) -> str:
    """The line that hands the reader to the next act.

    A scroll only works if each scene asks for the next one, so the handoff is
    part of the frame rather than something each act remembers to write. The
    last act has nothing to hand to.

    It is appended to the act body, not to the section around it: the profile
    switch replaces the whole body with the copy it stored, so a bridge added
    outside would survive the first paint and vanish on the first swap.
    """
    key = f"act.{act.id}.next"
    return "" if key not in STRINGS else f'<p class="act-next">{t(key)}</p>' 


def rail() -> str:
    """Thirteen acts in uppercase mono, grouped into the three parts."""
    out, part = "", None
    for act in ACTS:
        if act.part != part:
            out += "</ol>" if part is not None else ""
            out += (f'<p class="rail-part">{t(f"part.{act.part}")}</p><ol>')
            part = act.part
        out += (f'<li><a href="#act-{act.id}" data-rail="{act.id}">'
                f'<span class="num">{act.id}</span>{t(act.title)}</a></li>')
    return out + "</ol>"


def render(payload: dict, bundles: dict) -> str:
    """Returns the page. Part 2 is added to `payload` as a side effect."""
    default = payload["meta"]["profiles"][0]
    per_profile: dict[str, dict[str, str]] = {u: {} for u in payload["profiles"]}

    html = SHELL.read_text()
    for act in ACTS:
        if act.per_profile:
            for user in per_profile:
                ctx = Context(payload, bundles, user)
                per_profile[user][act.id] = act.builder(ctx) + bridge(act)
            body = per_profile[default][act.id]
        else:
            body = act.builder(Context(payload, bundles)) + bridge(act)
        html = html.replace(f"<!--act:{act.id}-->", section(act, body))

    # The default profile's part two is already in the document, so the page
    # keeps that copy rather than downloading it a second time.
    for user, acts in per_profile.items():
        payload["profiles"][user]["acts"] = {} if user == default else acts
    payload["meta"]["default_profile"] = default

    return (html
            .replace("<!--title-->", t("site.page_title"))
            .replace("<!--description-->", t("cover.standfirst"))
            .replace("<!--rail-->", rail())
            .replace("<!--pill-->",
                     f'{t("pill.label")} <span class="who">{default}</span>'))
